File: apps/cargo-tracker/tests/CargoTracker.Domain.Tests/Shipper/ShipperTest.cs
using CargoTracker.Shipper.Domain.Model;
using CargoTracker.Shipper.Domain.Model.Events;
using FluentAssertions;

namespace CargoTracker.Domain.Tests.Shipper;

public class ShipperTest
{
    [Fact]
    public void 個人荷主を登録でき割引率はゼロで荷主コードが自動採番される()
    {
        // When: 個人荷主を登録する
        var shipper = CargoTracker.Shipper.Domain.Model.Shipper.RegisterIndividual(
            new ShipperName("山田太郎"), new Email("yamada@example.com"), new Phone("03-1234-5678"));

        // Then
        shipper.Type.Should().Be(ShipperType.Individual);
        shipper.DiscountRate.Value.Should().Be(0m);
        shipper.ContractNumber.Should().BeNull();
        shipper.Code.Value.Should().StartWith("SHP-");
    }

    [Fact]
    public void 法人荷主を登録でき契約番号と割引率を保持する()
    {
        // When: 法人荷主を登録する
        var shipper = CargoTracker.Shipper.Domain.Model.Shipper.RegisterCorporate(
            new ShipperName("株式会社サンプル"), new Email("corp@example.com"), null,
            new ContractNumber("C-0001"), new DiscountRate(0.15m));

        // Then
        shipper.Type.Should().Be(ShipperType.Corporate);
        shipper.ContractNumber!.Value.Should().Be("C-0001");
        shipper.DiscountRate.Value.Should().Be(0.15m);
    }

    [Fact]
    public void 荷主登録時にドメインイベントが発生する()
    {
        var shipper = CargoTracker.Shipper.Domain.Model.Shipper.RegisterIndividual(
            new ShipperName("山田太郎"), new Email("yamada@example.com"), null);

        shipper.PullDomainEvents().Should().ContainSingle(e => e is ShipperRegisteredEvent);
    }

    [Theory]
    [InlineData(-0.0001)]
    [InlineData(0.3001)]
    [InlineData(1.0)]
    public void 割引率が0から30パーセントの範囲外なら例外(decimal invalidRate)
    {
        var act = () => new DiscountRate(invalidRate);
        act.Should().Throw<ArgumentOutOfRangeException>();
    }

    [Theory]
    [InlineData(0.0)]
    [InlineData(0.15)]
    [InlineData(0.30)]
    public void 割引率が0から30パーセントの範囲内なら生成できる(decimal validRate)
    {
        var rate = new DiscountRate(validRate);
        rate.Value.Should().Be(validRate);
    }

    [Fact]
    public void 不正なメールアドレスは例外()
    {
        var act = () => new Email("invalid-email");
        act.Should().Throw<ArgumentException>();
    }

    [Fact]
    public void 荷主名が空なら例外()
    {
        var act = () => new ShipperName("  ");
        act.Should().Throw<ArgumentException>();
    }
}

File: apps/cargo-tracker/src/contexts/shipper/domain/model/shipper.test.ts
import { describe, expect, it } from 'vitest';
import { Shipper } from './shipper.js';
import { ShipperType } from './value-objects.js';

describe('Shipper 集約', () => {
  it('個人荷主を登録できる（割引率ゼロ・契約番号なし）', () => {
    const shipper = Shipper.registerIndividual({
      name: '山田太郎',
      email: 'yamada@example.com',
    });
    expect(shipper.shipperType).toBe(ShipperType.INDIVIDUAL);
    expect(shipper.code.value).toMatch(/^SHP-/);
    expect(shipper.discountRate.value).toBe(0);
    expect(shipper.contractNumber).toBeUndefined();
  });

  it('法人荷主を契約番号・割引率つきで登録できる', () => {
    const shipper = Shipper.registerCorporate({
      name: '株式会社サンプル',
      email: 'corp@example.com',
      contractNumber: 'CT-001',
      discountRate: 0.2,
    });
    expect(shipper.shipperType).toBe(ShipperType.CORPORATE);
    expect(shipper.contractNumber?.value).toBe('CT-001');
    expect(shipper.discountRate.value).toBe(0.2);
  });

  it('法人で割引率が範囲外なら登録できない', () => {
    expect(() =>
      Shipper.registerCorporate({
        name: '株式会社サンプル',
        email: 'corp@example.com',
        contractNumber: 'CT-001',
        discountRate: 0.5,
      }),
    ).toThrow();
  });

  it('永続化値から再構築できる', () => {
    const shipper = Shipper.reconstruct({
      id: 10,
      code: 'SHP-abcd1234',
      shipperType: ShipperType.CORPORATE,
      name: '株式会社サンプル',
      email: 'corp@example.com',
      phone: null,
      address: null,
      contractNumber: 'CT-001',
      discountRate: 0.1,
    });
    expect(shipper.id).toBe(10);
    expect(shipper.contractNumber?.value).toBe('CT-001');
  });
});
